### ISSUE TITLE
Is there an example to follow? 

### ISSUE THREAD

**Issue by charlespedro** [JHU-DataViz-Summer16/week-4-replication-project#9]
I looked for an example to get going here. As usual, I found a Mike example but he's using v4, and our current v3 script won't work that way. 

https://bl.ocks.org/mbostock/3048450

When I try to replicate what he does using v3, console says "d3.histogram is not a function"

```
var bins = d3.histogram()
    .domain(x.domain())
    .thresholds(x.ticks(20))
    (data[4]);
```

Also, v4 uses scaleLinear where we use scale.linear. So....I don't really have a clue what to do here.

**Comment by rmcarder** [comment]
I changed this part to:

var bins = d3.layout.histogram()
      .bins(x.ticks().length-1)
      (data);

Since in version 4 the domain and thresholds properties replaced the bin property from version 3(according to HW readme). Adding layout got it to stop telling me histogram isnt a function. 

Still unsure about the next step, however. Read the documentation on the 3 attributes that each bin contains, and tried to do something similar to creating the scatterplots while also using Bostock's example to guide. Tried to change the x0, x0 notation into the dx,x,and y from version 3 but width is coming back as NaN:

var bins = d3.layout.histogram()
      .bins(x.ticks().length-1)
      (txData);

```
var bar = svg.selectAll('.bar')
  .data(bins)
  .enter().append('g')
  .attr('class', 'bar')
  .attr("transform", function(d) { return "translate(" + x(d.x) + "," + y(d.y) + ")"; });

bar.append('rect')
  .attr('width', x(bins[0].dx))
  .attr('height', function (d) {return height - y(d.y); });
}
```

**Comment by mike-bales** [comment]
I was finally able to get the histogram to work. The key for me was using the `x.ticks()` function which returns the values for your tick marks. If you just run it in the console window in chrome you will see what it returns which is very helpful. 

So in Rich's code he is feeding `x.ticks().length-1` as the number of bins. You can also just straight feed it `x.ticks()` since this will return an array of the value of your tick marks which are also the thresholds for the bins we want to create since our bins correspond to the area between the tick marks on the graph.

``` javascript
var bins = d3.layout.histogram()
.bins(x.ticks().length-1)
(txData);
```

I think where you are getting tripped up is in the data you are feeding it. txData is all of your data but you only want to bin (and histogram only accepts) one variable at time. You can use `histogram.value` to get that one variable or you can create a variable with the data you need and feed that to `layout.histogram`. I did the latter using `.map`. Check out [this youtube](https://www.youtube.com/watch?v=cu-I2um024k) for where I got this .

``` javascript
// HISTOGRAMS

    var demContr = txData.map(function (i) {return i.democratContribution});

    var myHistogram = d3.layout.histogram()
                             .bins(x.ticks())
                             (demContr);
```

If you look at what comes out of `layout.histogram` using `console.log(myHistogram)` you see that its just an array. Once you understand the structure of the array that comes out of `layout.histogram` its easier to understand how to turn that into a visualization. Check out the video I linked above (its only 11 minutes) for a walk through of the array structure.

The next thing I did was create a scale to translate the data in the histogram array to the scale of my svg. I was having trouble understanding scales and axis but [this post](http://alignedleft.com/tutorials/d3/scales) really helped. The key paragraph for me was:

> It may be tempting to think of a scale as something that appears visually in the final image — like a set of tick marks, indicating a progression of values. Do not be fooled! Those tick marks are part of an axis, which is essentially a visual representation of a scale. A scale is a mathematical relationship, with no direct visual output. I encourage you to think of scales and axes as two different, yet related, elements.

Here is my scale:

``` javascript
    histScale = d3.scale.linear()
                        .domain([0, d3.max(histogram, function (d) { return d.y; })]) //used histogram array to get max bucket value
                        .range([0, height/5]) //set output range to 1/5 the height of the existing graph. 
                        .nice();
//In essence my largest bar will now be 1/5 the height of the graph
```

Now I can use that scale to set the height of my bars. 

``` javascript
// Bind histogram array to g elements that I tag with bar
  var bars = svg.selectAll(".bar")
                  .data(histogram)
                  .enter().append("g")
                  .attr("class", "bar");

  bars.append("rect")
         .attr("width", 25)
         .attr("height", function (d) { return histScale(d.y)}); //feed d.y (an element from the histogram array  that has the number of values in that bin) to histScale to get the scaled height in pixels for each bin
```

Now this will just give me a number of rects of width 25 and height scaled to the value of that bin. They will all sit on top of one another. So I need to position them.

Here is where the understanding of scales as function that just translates values comes in handy. If I feed my x scale any of my tick values it will give me back the pixel position of that tick. In the console if you enter `x(x.ticks()[1])` you are feeding the x scale the second value from the array you get back from `x.ticks()` (the second because the first is 0). This will result in the pixel position for that value. Now that you have this you can create a function that finds the center point of each column as you run through the histogram array when creating your rects. Here is my final code for positioning the rects

``` javascript

// Bind histogram array to g elements that I tag with bar
  var bars = svg.selectAll(".bar")
                  .data(histogram)
                  .enter().append("g")
                  .attr("class", "bar");

// x axis histogram

var barWidth = ( x(x.ticks()[1]) - x(x.ticks()[0]) )/5; //set width of bars to 1/5 of distance between the first two ticks

bars.attr("transform", function(d, i) {
                       return "translate(" + 
                          (  //Position bars across top x axis
                               ((x(x.ticks()[i+1]) - (x(x.ticks()[i])))/2) 
                               + x(x.ticks()[i]) 
                                - (barWidth/2) 
                          )  
                             + ",0)"; 
                 }) 
       .append("rect")
       .attr("width", barWidth)
       .attr("height", function (d) { return histScale(d.y)});
```

**Comment by charlespedro** [comment]
Yeah thanks for the explanation. I should have responded sooner - I got the x-axis histogram working this morning but wasn't sure how to center the bars neatly between the tick marks without hardcoding it. 

Your barWidth solution is a good one. Cheers!

**Comment by cmgiven** [comment]
Apparently I wasn't watching this repo, so I wasn't notified of this issue being posted. Great discussion, though.

Three notes on the last code snippet Mike posted for anyone following along:
- Note that because we immediately call `.enter()` after joining the data, the `bars` variable contains just the enter selection. This means that once we have added all of the bars the first time `update()` is called, we'll never change them again — there's always the same number of bars, so nothing to enter, thus all of this code is called on an empty selection the second and subsequent times it runs. We would instead call `bars.enter()...` on a new line so that we can continue to access the full `bars` selection.
- As we showed in class, we can use `dx` from the histogram layout to figure out the total width between two ticks. That could also help simplify some of the math a bit.
- As we noted in class, if we only want to add a single element per datum, there's really no reason to use a group. We can just add the rect directly.
